File: wrappers/rust/src/barcode_format.rs
use crate::bindings;
use flagset::flags;
use num_enum::FromPrimitive;
use std::fmt::{Display, Formatter};

flags! {
    #[repr(u32)]
    #[derive(FromPrimitive)]
    pub enum BarcodeFormat: u32 {
        #[num_enum(default)]
        None = 0,
        /// Aztec
        Aztec = 1 << 0,
        /// Codabar
        Codabar = 1 << 1,
        /// Code39
        Code39 = 1 << 2,
        /// Code93
        Code93 = 1 << 3,
        /// Code128
        Code128 = 1 << 4,
        /// GS1 DataBar, formerly known as RSS 14
        DataBar = 1 << 5,
        /// GS1 DataBar Expanded, formerly known as RSS EXPANDED
        DataBarExpanded = 1 << 6,
        /// DataMatrix
        DataMatrix = 1 << 7,
        /// EAN-8
        EAN8 = 1 << 8,
        /// EAN-13
        EAN13 = 1 << 9,
        /// ITF (Interleaved Two of Five)
        ITF = 1 << 10,
        /// MaxiCode
        MaxiCode = 1 << 11,
        /// PDF417
        PDF417 = 1 << 12,
        /// QR Code
        QRCode = 1 << 13,
        /// UPC-A
        UPCA = 1 << 14,
        /// UPC-E
        UPCE = 1 << 15,
        /// Micro QR Code
        MicroQRCode = 1 << 16,
        /// Rectangular Micro QR Code
        RMQRCode = 1 << 17,
        /// DX Film Edge Barcode
        DXFilmEdge = 1 << 18,

        LinearCodes = (BarcodeFormat::Codabar | BarcodeFormat::Code39 | BarcodeFormat::Code93 | BarcodeFormat::Code128 | BarcodeFormat::EAN8 | BarcodeFormat::EAN13 | BarcodeFormat::ITF | BarcodeFormat::DataBar | BarcodeFormat::DataBarExpanded | BarcodeFormat::DXFilmEdge | BarcodeFormat::UPCA | BarcodeFormat::UPCE).bits(),
        MatrixCodes = (BarcodeFormat::Aztec | BarcodeFormat::DataMatrix | BarcodeFormat::MaxiCode | BarcodeFormat::PDF417 | BarcodeFormat::QRCode | BarcodeFormat::MicroQRCode | BarcodeFormat::RMQRCode).bits(),
        Any = (BarcodeFormat::LinearCodes | BarcodeFormat::MatrixCodes).bits(),
    }
}

impl Display for BarcodeFormat {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            bindings::ffi::barcode_format_to_string(*self as i32)
        )
    }
}

File: wrappers/kn/src/nativeMain/kotlin/zxing/BarcodeFormat.kt
package zxing

import kotlinx.cinterop.ExperimentalForeignApi
import zxing.cinterop.*

@OptIn(ExperimentalForeignApi::class)
enum class BarcodeFormat(internal val rawValue: UInt) {
	None(zxing_BarcodeFormat_None),
	Aztec(zxing_BarcodeFormat_Aztec),
	Codabar(zxing_BarcodeFormat_Codabar),
	Code39(zxing_BarcodeFormat_Code39),
	Code93(zxing_BarcodeFormat_Code93),
	Code128(zxing_BarcodeFormat_Code128),
	DataBar(zxing_BarcodeFormat_DataBar),
	DataBarExpanded(zxing_BarcodeFormat_DataBarExpanded),
	DataMatrix(zxing_BarcodeFormat_DataMatrix),
	DXFilmEdge(zxing_BarcodeFormat_DXFilmEdge),
	EAN8(zxing_BarcodeFormat_EAN8),
	EAN13(zxing_BarcodeFormat_EAN13),
	ITF(zxing_BarcodeFormat_ITF),
	MaxiCode(zxing_BarcodeFormat_MaxiCode),
	PDF417(zxing_BarcodeFormat_PDF417),
	QRCode(zxing_BarcodeFormat_QRCode),
	MicroQrCode(zxing_BarcodeFormat_MicroQRCode),
	RMQRCode(zxing_BarcodeFormat_RMQRCode),
	UPCA(zxing_BarcodeFormat_UPCA),
	UPCE(zxing_BarcodeFormat_UPCE),

	LinearCodes(zxing_BarcodeFormat_LinearCodes),
	MatrixCodes(zxing_BarcodeFormat_MatrixCodes),
	Any(zxing_BarcodeFormat_Any),

	Invalid(zxing_BarcodeFormat_Invalid),
}

@OptIn(ExperimentalForeignApi::class)
fun zxing_BarcodeFormat.parseIntoBarcodeFormat(): Set<BarcodeFormat> =
	BarcodeFormat.entries.filter { this.or(it.rawValue) == this }.toSet()

@OptIn(ExperimentalForeignApi::class)
fun Iterable<BarcodeFormat>.toValue(): zxing_BarcodeFormat =
	this.map { it.rawValue }.reduce { acc, format -> acc.or(format) }
